[Commands] Add opt. global `[<TaskName>].` prefix for all PLUGIN_WRITE commands

tonhuisman opened the pull request:
Support for `[<TaskName>].` and `[<TaskNumber>].` command-prefix made global. Assuming that the plugin supports multiple instances, that's possibly limited by the plugin implementation or plugin nature.
Includes documentation.

@TD-er : 
- what to do with P022 PCA9685 ~~possibly buggy~~ quite specific implementation? (excluded for now, best leave it this way)
- Does `irsendac` (P035), used as an example, actually support multiple instance?

Closes #3195 and closes #3196

tonhuisman committed:
[Commands] Add global `[<TaskName>].` prefix for all PLUGIN_WRITE commands

tonhuisman committed:
[Commands] Remove local taskname-prefix impl., exclude P022

tonhuisman commented:
- Removed the local `[<TaskName>].` prefix implementations for P023, P095 and P096
- Made explicit exclusion for P022, as that plugin has very specific handling when addressed with the `[<TaskName>].` prefix, decided to leave that as-is, for 100% compatibility.

tonhuisman committed:
[Commands] Remove requirement for [ and ] on <TaskName>. prefix

tonhuisman commented:
- Made the implementation more backward compatible, no longer requiring the use of square brackets around `<TaskName>`
- Minor code-cleanup for P022

tonhuisman committed:
[Commands] Improve documentation on [<TaskName>]. prefixing

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
[Commands] Add extra safeguard check for DeviceIndex

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman commented:
Only improvement I can think of here is to check that when square braces are used, they are the first and last character of the taskname, but that sounds just a bit superfluous, as you can't use square braces in tasknames anyway.

@TD-er , any improvement suggestions, or is this one good to go? (I'd need this to be merged for new SPI plugins I'm still working on, to hatch properly 😸) (Adding reviewers doesn't seem to work anymore in the new github UI?)

TD-er commented:
This is something I have to look into when I'm able to concentrate (when my daughter is to bed and indoor temperature is in acceptable range)

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
[P023] Fix merge conflict

TD-er commented on src/__Plugin.ino:
```diff
@@ -1200,14 +1200,46 @@ byte PluginCall(byte Function, struct EventStruct *event, String& str)
+        }
+      }
+
+      for (taskIndex_t task = firstTask; task < lastTask; task++)
```
The inner scope of this loop should be in another function.
In this loop you're trying to determine whether we need to "find" the task willing to accept the command (needs the for loop)
The added feature to directly address a task either by name or by index (not yet implemented) is to extract a specific task index.
So by having a separate function to perform this on a single task index you improve readability of the code.

tonhuisman replied:
And that function should then also be called from the other PluginCall functions, I assume? They are *mostly* similar, but not exactly the same.

tonhuisman replied:
@TD-er please review my refactored code, as C/C++ isn't my daily bread and butter... (just dabbling here 😛).

TD-er commented on src/__Plugin.ino:
```diff
@@ -1200,14 +1200,46 @@ byte PluginCall(byte Function, struct EventStruct *event, String& str)
+      int dotPos = command.indexOf(".");                    // Find first period
+      if (Function == PLUGIN_WRITE                          // Only applicable on PLUGIN_WRITE function
+        && dotPos > -1) {                                   // First precondition is just a quick check for a period (fail-fast strategy)
+        String thisTaskName = command.substring(1, dotPos); // Extract taskname
```
This is not correct.
The parameter separater we always used is still valid here.
`[name].command<comma>param1...`
So please use the `parseString` to find the proper substring first.  (is in StringConverter.cpp)
On this substring (the first argument, sadly not counting from 0) is the command with the optional task index information.

On this substring you can test for a dot and try to find the task index from that information.
If you have the task index, call the function I describe in a comment below.

tonhuisman replied:
Fixed

TD-er commented on src/__Plugin.ino:
```diff
@@ -1200,14 +1200,46 @@ byte PluginCall(byte Function, struct EventStruct *event, String& str)
+                thisTask = INVALID_TASK_INDEX;
+              }
+            }
+            if (thisTask != INVALID_TASK_INDEX) {
```
Please use `validTaskIndex`

tonhuisman replied:
Fixed

TD-er commented on src/__Plugin.ino:
```diff
@@ -1200,14 +1200,46 @@ byte PluginCall(byte Function, struct EventStruct *event, String& str)
+        thisTaskName.replace("[", "");
+        thisTaskName.replace("]", "");
+        if (thisTaskName.length() > 0) {                    // Second precondition
+          taskIndex_t thisTask = findTaskIndexByName(thisTaskName);
```
Maybe also nice to add an extra check, if the task name is not found.
For example I know some users keep using duplicate task names and they may want to use an integer for the task index (`1 ... TASK_MAX` instead of `0 ... TASK_MAX-1`)
So `if (!validTaskIndex(thisTask))` try to see if it is a valid numerical and then see if that can be translated into a valid task index.

tonhuisman replied:
Added as a feature 👍

uzi18 commented:
@TD-er @tonhuisman it was discussed somewhere in issues/PR about addressing specific task for command execution
@TD-er showed also idea about addressing external unit, maybe we should take some steps back and verify it to have proper & readable syntax
Or maybe have missed something I'm more offline from couple of months ;)

tonhuisman commented:
@uzi18 I haven't seen or found that discussion (about addressing an external unit) (yet), but I'm curious about any details available.

OTOH:
Couldn't that be handled by using the `SendTo <unit>,<command>` command?
Though `sendto` does require the ESPEasy UDP network being configured, it is AFAICS the only way one unit can address another unit without knowing the exact IP address. And you still have to know the unit number (or name?).

tonhuisman committed:
[Commands] Improvements, refactoring and added feature of [<TaskNumber>]. prefix

tonhuisman commented:
Improvements and suggestions applied
Added feature of prefixing with `[<TaskNumber>]` if `[<TaskName>]` can not be found.
Refactored partially, last bits to be discussed/reviewed if that should be refactored, as it doesn't help for code-clarity (but added an `unConditional` input argument already).

tonhuisman committed:
[Docs] Updated for [<TaskNumber>]. prefix

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/explicit_task_command

tonhuisman commented:
A manual merge was needed because of the many improvements that where made the last couple of weeks.

tonhuisman committed:
[Checks] Don't allow numeric (float) device name

tonhuisman commented:
NB: The checks are inhibited by the `# define LIMIT_BUILD_SIZE`, applied to some configurations (a.o. test), maybe this code should be left in. It does *cost* around 900 bytes of flash space.

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/explicit_task_command

tonhuisman committed:
[TaskName] Define is no longer available

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/explicit_task_command

TD-er commented on src/src/Globals/Plugins.cpp:
```diff
@@ -243,24 +243,26 @@ void post_I2C_by_taskIndex(taskIndex_t taskIndex, deviceIndex_t DeviceIndex) {
+        TempEvent->setTaskIndex(taskIndex);
         TempEvent->sensorType   = Device[DeviceIndex].VType;
+        if (event != nullptr) {
+          TempEvent->OriginTaskIndex = event->TaskIndex;
```
Why do you need the OriginTaskIndex?

tonhuisman replied:
I tried to merge over the changes you made, don't know if it is really needed. If not I can remove it.

TD-er replied:
Ah OK, if it is indeed a change I made, then leave it :)

I only saw it as your changes, did not notice it was a rebase merge.

TD-er commented on src/src/Globals/Plugins.cpp:
```diff
@@ -243,24 +243,26 @@ void post_I2C_by_taskIndex(taskIndex_t taskIndex, deviceIndex_t DeviceIndex) {
  * Call the plugin of 1 task for 1 function, with standard EventStruct and optional command string
  */
-bool PluginCallForTask(taskIndex_t task, byte Function, EventStruct *TempEvent, String& command, bool unConditional = false) {
+bool PluginCallForTask(taskIndex_t taskIndex, byte Function, EventStruct *TempEvent, String& command, EventStruct *event = nullptr) {
```
You have access to the TempEvent, but is there any reason why the `event` should not be a `const EventStruct*` ?

Also the default assignment should only be done in a .h file, not in the .cpp implementation.

tonhuisman replied:
This isn't declared in the .h file, should only be called from PluginCall

TD-er replied:
I see.

tonhuisman commented:
This is a change that is still not decided upon whether to keep or drop, AFAICS. Would be nice to get to a decision, so I can stop the manual merge every time Plugins.cpp is modified 😉

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

TD-er committed:
Merge branch 'mega' into feature/explicit_task_command

TD-er committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'feature/explicit_task_command' of https://github.com/tonhuisman/ESPEasy-1 into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'feature/explicit_task_command' of https://github.com/tonhuisman/ESPEasy-1 into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'feature/explicit_task_command' of https://github.com/tonhuisman/ESPEasy-1 into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

TD-er committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command

tonhuisman committed:
Merge branch 'mega' into feature/explicit_task_command
